### PR TITLE
React flow integration

### LINKED ISSUE #956: [FEATURE]: Integrate React Flow
### Feature Name

Integrate React Flow

### Your Name

Faris Durrani

### Description

Replace the React-DnD component with a React Flow component that will allow multi-layer neural networks

### PR DESCRIPTION
**Pull request by farisdurrani**
# React flow integration

**What user problem are we solving?**
The current ReactDnD tool is complex in nature and does not allow for parallel layers of architecture.

**What solution does this PR provide?**
Addressing #890 , this PR integrates React Flow as an alternative drag-n-drop tool replacing ReactDnd. This allows people to drag and drop new layers, as well as add parameter info and in the future, this will allow parallel layers of architecture.

https://github.com/DSGT-DLP/Deep-Learning-Playground/assets/40067313/1130fa24-a345-4ea2-a092-f200bdf2808f

**Testing Methodology**
How did you test your changes and verify that existing 
functionality is not broken

Basic testing on Iris and MNIST training.

**Any other considerations**

### CODE REVIEW

**Comment by karkir0003** [comment]
@farisdurrani screen recording?

File: .github/Architecture.md [hunk 1]
```diff
@@ -77,6 +77,7 @@
 |  |  |  |  |  |  |- 📂 constants:
 |  |  |  |  |  |  |  |- 📜 imageConstants.ts
 |  |  |  |  |  |  |- 📂 components:
+|  |  |  |  |  |  |  |- 📜 ImageFlow.tsx
```
Comment (karkir0003) [inline .github/Architecture.md]: doesn't the file tree generator script build this for us?

File: .vscode/settings.json [hunk 2]
```diff
@@ -19,5 +19,6 @@
   },
   "python.formatting.provider": "none",
   "python.terminal.activateEnvironment": false,
-  "python.analysis.typeCheckingMode": "basic"
+  "python.analysis.typeCheckingMode": "basic",
+  "cSpell.words": ["reactflow"]
```
Comment (karkir0003) [inline .vscode/settings.json]: explain? confused

File: README.md [hunk 3]
```diff
@@ -177,3 +177,4 @@ See [Architecture.md](./.github/Architecture.md)
 Deep Learning Playground is MIT licensed, as found in the [LICENSE](./LICENSE) file.
 
 Deep Learning Playground documentation is Creative Commons licensed, as found in the [LICENSE-docs](./.github/LICENSE-docs) file.
+
```
Comment (karkir0003) [inline README.md]: nit: extra newline

File: frontend/src/features/Train/features/Image/components/ImageFlow.tsx [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+    ...DEFAULT_LAYERS.MNIST.map((layer, i) => ({
+      id: `${layer.value}-${i}`,
+      type: "textUpdater",
+      position: {
+        x: DEFAULT_X_POSITION,
+        y: (i + 1) * 100,
+      },
+      data: {
+        label: STEP_SETTINGS.PARAMETERS.layers[layer.value].label,
+        value: layer.value,
```
Comment (karkir0003) [inline frontend/src/features/Train/features/Image/components/ImageFlow.tsx:55]: can you show screen recording of image flow. right now i see mainly tabular flow

File: frontend/src/features/Train/features/Image/components/ImageFlow.tsx [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+    })),
+  ];
+  const initialEdges: Edge[] = createInitialEdges();
+
+  const [nodes, setNodes, onNodesChange] = useNodesState(initialNodes);
+  const [edges, setEdges, onEdgesChange] = useEdgesState(initialEdges);
+
+  useEffect(() => {
+    if (setLayers) {
+      const layers = handleExportLayers();
```
Comment (karkir0003) [inline frontend/src/features/Train/features/Image/components/ImageFlow.tsx:68]: could this theoretically be slow if we had a humongous model?

non-blocking, but a question

**Review by karkir0003 — COMMENTED** [review]
looks good to me. left a few comments/questions.

@dwu359 , can you take a look

File: .github/Architecture.md (hunk 1 above)
Comment (farisdurrani) [reply .github/Architecture.md]: Yes! This is automatically generated

File: .vscode/settings.json (hunk 2 above)
Comment (farisdurrani) [reply .vscode/settings.json]: Sorry, this isn't meant to be uploaded. In fact, all of .vscode should be gitignored

Comment (farisdurrani) [reply .vscode/settings.json]: If you're wondering, I'm using Code Spell Checker and 'reactflow' isn't recognized as an English word, so I added it to my custom dictionary

File: README.md (hunk 3 above)
Comment (farisdurrani) [reply README.md]: good catch

File: frontend/src/features/Train/features/Image/components/ImageFlow.tsx (hunk 4 above)
Comment (farisdurrani) [reply frontend/src/features/Train/features/Image/components/ImageFlow.tsx:55]: Summary edited to include screen recording

File: frontend/src/features/Train/features/Image/components/ImageFlow.tsx (hunk 5 above)
Comment (farisdurrani) [reply frontend/src/features/Train/features/Image/components/ImageFlow.tsx:68]: Theoretically, yes. The way how it works is the layers state get updated at every node or edge change. It is an expensive operation but so far, it's non-significant for the number of layers we have rn

**Review by dwu359 — APPROVED** [review]
This looks really good! Future steps could be transforming the entire trainspace ui to use react flow instead of just the layers.

**Pull request merged**